Get SQL databases working again

TheLimeGlass opened the pull request:
### Description
- Adds H2 as a configurable database. Will probably make default eventually or maybe in this pull request?
- Removes SQLibrary and adds https://github.com/brettwooldridge/HikariCP as the main SQL wrapper.
- Revamped API for better registration and handling of JDBC Java drivers.
- Added a `commit changes` option which is essentially how Skript used to operate. Changed the JDBC standard to auto commit after every edit (https://github.com/SkriptLang/Skript/issues/2007). This is standard and better for performance. Skript didn't for some reason. I speculate it was to allow MySQL to sync with other servers, so either way it's an option now.

### Perks
- Better configuration options with HikariCP.
- Way better performance with HikariCP.
- Easier implementation and not relying on a small project like SQLibary.
- Thread pooling potential with HikariCP.
- This new API design allows for quite literally any JDBC implementation database.
- Uses the SpigotLibraryLoader to load the resources at runtime rather than shadowing into the jar.

### TODO
- [x] Test MySQL.
- [x] Test SQLite. 
- [x] Test H2.
- [x] Run beta testing in the SkriptLang Discord.
- [x] Adds tests.
- [x] In H2 use H2's `MERGE INTO` instead of `INSERT`.
- [x] Forgot to add a check that HikariCP classes were loaded by SpigotLibraryLoader.
- [x] Rename SQLStorage to something with JDBC reference, maybe JdbcStorage as it's not only SQL anymore.
- [ ] Test that monitor changes still do something.
- [x] Attach a SkriptAddon instance to the registration, so Skript knows where a storage is coming from.
- [ ] Investigate if this solves https://github.com/SkriptLang/Skript/issues/2022

Optional
- [ ] Write up an API tutorial for registering custom storage types.
- [ ] Document all the possible options for databases or have annotations for a documentation system.
- [ ] See if Skript could handle async and multithreading now that the database has the possibility.
- [ ] Maybe add MongoDB support https://github.com/SkriptLang/Skript/issues/1787
- [ ] Change the existing behavior of dumping every variable into ram on server start.

### Notes:
- Should we just remove SQLite support? It's not like anyone is using it right now, it doesn't work currently.
- H2 is a flat file SQL implementation that supports asynchronous operations and multi threaded unlike SQLite in both those regards. 
- This does not impact anything to do with the default flat file CSV.
- This mainly only affects JDBC driver databases. The rework is still a viable experiment started by bensku https://github.com/SkriptLang/Skript/tree/feature/variables2

### Testing and using this jar
To test this experimental feature out, go to the "checks" tab and then click a Java version on the side and then click the nightly artifacts to get a built jar of the latest commit.

---
**Target Minecraft Versions:** any
**Requirements:** none
**Related Issues:** https://github.com/SkriptLang/Skript/issues/1168, https://github.com/SkriptLang/Skript/issues/2007, https://github.com/SkriptLang/Skript/issues/1478 and https://github.com/SkriptLang/Skript/issues/3948

TheLimeGlass committed:
Progress so far

TheLimeGlass committed:
Merge branch 'master' of https://github.com/SkriptLang/Skript into enhancement/sql

TheLimeGlass committed:
Finalize SQL database system

TheLimeGlass committed:
Remove comments

TheLimeGlass committed:
Remove SQLibrary

TheLimeGlass committed:
Complete SQL rework

TheLimeGlass committed:
Formatting

TheLimeGlass committed:
Tackle some todo list nodes

TheLimeGlass committed:
Class check

APickledWalrus reviewed (changes requested):
With a quick initial review of the current progress, these changes look excellent. Fantastic work! I would definitely like a review from others too as I am not the most familiar with SQL/the libraries being used

TheLimeGlass commented on src/main/java/ch/njol/skript/variables/SerializedVariable.java:
```diff
@@ -29,15 +29,15 @@ public class SerializedVariable {
 	 * The name of the variable.
 	 */
-	public final String name;
+	private final String name;
```
SerializedVariable should only have been used internally. I enforced this to be using the methods as it's now going to be accessable via registering a custom storage with the API.

TheLimeGlass commented on src/main/java/ch/njol/skript/variables/SQLiteStorage.java:
```diff
@@ -19,37 +19,92 @@
+	@Override
+	protected File getFile(String file) {
+		if (!file.endsWith(".db"))
+			file = file + ".db"; // required by SQLibrary
```
It's required by SQLite not SQLibrary, should update that comment. It was an existing broad statement.

xartuu commented:
Wow, am I dreaming? Awesome that someone is taking on the SQL problem!

Moderocky reviewed (commented):
The code looks fine, but I know very little about SQL so I don't feel I'm the right person to be reviewing this.

TheLimeGlass commented:
Not working on this anymore

TheLimeGlass committed:
Update to dev/feature

TheLimeGlass committed:
Apply changes

TheLimeGlass committed:
changes

TheLimeGlass committed:
Add SkriptAddon reference to the storage

TheLimeGlass committed:
Add SQLite tests

TheLimeGlass committed:
Fix MySQL

TheLimeGlass committed:
licence

TheLimeGlass committed:
Merge branch 'dev/feature' into enhancement/sql

TheLimeGlass committed:
Merge branch 'dev/feature' into enhancement/sql

TheLimeGlass committed:
Merge branch 'dev/feature' into enhancement/sql

TheLimeGlass committed:
Complete MySQL

xartuu commented:
Hello, do you have an implementation date already planned?

Moderocky commented:
> Hello, do you have an implementation date already planned?

It needs people to review it first.

Moderocky committed:
Merge branch 'dev/feature' into enhancement/sql

Lennord commented:
Good evening, I tested the latest build of your new SQL and I have these errors (below), when I activate auto-commit + put the minotor for around 1 seconds, I tried to leave it by default but i have the same errors.

Link File Log and view Error: [latest.log](https://github.com/SkriptLang/Skript/files/14762309/latest.log)

TheLimeGlass committed:
Merge branch 'dev/feature' into enhancement/sql

TheLimeGlass committed:
Fix java21 folder reference'

TheLimeGlass committed:
Better get method design. Monitor changes fixes. SQL tweaks

TheLimeGlass committed:
Undo rowid SQLite test

TheLimeGlass commented:
- Swapped method names load_i and load (The internal method was not load_i when it should have been)
- Potentially fixes monitor changes, needs testing on each database type that uses SQL.
- Changed the get method to be a Function rather than BiFunction and made it have a boolean if it's a test operation and also return the rowid for monitor changes which is required for MySQL and Jdbc database that wants monitor changes.
- Added loadJdbcConfiguration(SectionNode section) method so Jdbc extended databases can have their own custom configuration in the config.sk
- Fixed an issue where if the monitor changes queue rowid was not being set. This was due to monitor changes being enabled requiring the the auto commit to be set to false, so it now ensures commits for the rowid setting are set in the prepared statement.
- Changed the 'cannot load variables fast enough' error to not error on first load, as the first load will be more resource heavy. Long time coming TODO comment.
- Monitor changes now only work when a rowid has been returned when loading.

Test jar available when checks are complete at https://github.com/SkriptLang/Skript/actions/runs/9070821309

TheLimeGlass committed:
Merge branch 'dev/feature' into enhancement/sql

Prorickey commented:
What's being waited on to merge this?

EquipableMC commented:
> What's being waited on to merge this?

There has to be a lot of testing done before this can be merged, considering this is a big change, it will most likely come in 2.10.

TheLimeGlass committed:
Merge branch 'dev/feature' into enhancement/sql

Hundred-Killer commented:
If they fix it, I'll be happy.

TheLimeGlass committed:
Conflicts

TheLimeGlass committed:
Conflicts

TheLimeGlass committed:
Code cleaning and security

TheLimeGlass committed:
Code cleaning and security

TheLimeGlass commented on src/main/resources/plugin.yml:
```diff
@@ -31,6 +31,10 @@ main: ch.njol.skript.Skript
 version: @version@
 api-version: 1.13
 
+libraries:
```
dev/feature no longer supports 1.16, marking resolved.

TheLimeGlass committed:
Update Hikari and H2

TheLimeGlass committed:
Merge branch 'dev/feature' into enhancement/sql

TheLimeGlass committed:
git conflicts

TheLimeGlass committed:
Aliases

TheLimeGlass committed:
More git conflicts

sovdeeth commented on src/main/java/ch/njol/skript/variables/Variables.java:
```diff
@@ -237,22 +248,29 @@ public static boolean load() {
+					if (variablesStorage instanceof SQLiteStorage) {
+						Skript.warning(
+								"Please be advised that SQL is single threaded and synchronous.\n" +
+								"Meaning you will be losing out on drastic performance.\n" +
```
```suggestion
								"Meaning you will be losing out on drastic performance.\n" +
```
drastic performance what?
i think this should say `drastic performance improvements`

sovdeeth reviewed (approved):
The skript side of this PR looks solid. I am not familiar enough with SQL to comment on the storage implementations, but I'll be testing it to make sure it works.

Only a few small comments (the licenses at the start of files should also be removed)

nice work lime!

sovdeeth commented:
Also some pretty big conflicts that need resolving, and I unfortunately am not familiar enough with the code to do it for ya

APickledWalrus commented on src/main/java/ch/njol/skript/variables/SerializedVariable.java:
For the modified fields/added methods, `@Nullable` annotations can be placed inline before the type.

APickledWalrus commented on src/main/java/ch/njol/skript/variables/VariableStorage.java:
```diff
@@ -172,12 +215,13 @@ protected <T> T getValue(SectionNode sectionNode, String key, Class<T> type) {
 	 * @return whether the loading succeeded.
 	 */
-	public final boolean load(SectionNode sectionNode) {
+	public final boolean load_i(SectionNode sectionNode) {
```
Not sure about a public method being suffixed `_i`. Could we come up with a different name?

APickledWalrus commented on src/main/java/org/skriptlang/skript/variables/storage/SQLiteStorage.java:
```diff
@@ -0,0 +1,112 @@
+import org.jetbrains.annotations.ApiStatus.ScheduledForRemoval;
+import org.jetbrains.annotations.Nullable;
+
+@Deprecated
```
`@deprecated` javadocs would be good to point users to H2/MySQL storage

APickledWalrus commented on src/main/java/ch/njol/skript/variables/JdbcStorage.java:
```diff
@@ -0,0 +1,557 @@
+import ch.njol.util.NonNullPair;
+import ch.njol.util.SynchronizedReference;
+
+public abstract class JdbcStorage extends VariableStorage {
```
docs?

APickledWalrus commented on src/main/java/ch/njol/skript/variables/JdbcStorage.java:
```diff
@@ -0,0 +1,557 @@
+			try {
+				Connection connection = database.getConnection();
+				try {
+					if (WRITE_QUERY != null)
```
Can the `try` statements go within the null checks?

APickledWalrus commented on src/main/java/ch/njol/skript/variables/JdbcStorage.java:
```diff
@@ -0,0 +1,557 @@
+	@Override
+	protected final boolean loadAbstract(SectionNode section) {
+		synchronized (database) {
+			Timespan monitor_interval = getValue(section, "monitor interval", Timespan.class);
```
I would prefer to stick to camel case

APickledWalrus reviewed (changes requested):
Looks pretty good. My SQL knowledge is limited, but from what I do know, it seems fine. I know this has gone through testing on Discord as well.
Just a few minor things, re-request for approval

Lennord commented:
@TheLimeGlass Hi, can you update for Skript 2.10.1 and fixed the big issues with auto-commit please.

Lennord commented:
@TheLimeGlass Hello, can you fix the issues reported on auto-commit please?

sovdeeth commented:
closing in favor of https://github.com/SkriptLang/Skript/pull/7653
